Modules/itertoolsmodule.c

```diff
@@ -51,6 +53,169 @@ static PyTypeObject pairwise_type;
+   and there is no easy way for a user to detect that condition.
+   The same reasoning does not apply to batches() which never drops
+   data.  Instead, it produces a shorter list which can be handled
+   as the user sees fit.
+ */
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t batch_size;
+    PyObject *it;
+} batchedobject;
+
+/*[clinic input]
+@classmethod
+itertools.batched.__new__ as batched_new
+    iterable: object
+    n: Py_ssize_t
+Batch data into lists of length n. The last batch may be shorter than n.
+
+Loops over the input iterable and accumulates data into lists
+up to size n.  The input is consumed lazily, just enough to
+fill a list.  The result is yielded as soon as a batch is full
+or when the input iterable is exhausted.
+
+    >>> for batch in batched('ABCDEFG', 3):
+    ...     print(batch)
+    ...
+    ['A', 'B', 'C']
+    ['D', 'E', 'F']
+    ['G']
+
+[clinic start generated code]*/
+
+static PyObject *
+batched_new_impl(PyTypeObject *type, PyObject *iterable, Py_ssize_t n)
+/*[clinic end generated code: output=7ebc954d655371b6 input=f28fd12cb52365f0]*/
+{
+    PyObject *it;
+    batchedobject *bo;
+
+    /* XXX Do we want do defined batched(it, n=0) as just an empty iterator ?*/
+    if (n < 1) {
+        PyErr_SetString(PyExc_ValueError, "n must be >= 1");
+        return NULL;
+    }
+    it = PyObject_GetIter(iterable);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    /* create batchedobject structure */
+    bo = (batchedobject *)type->tp_alloc(type, 0);
+    if (bo == NULL) {
+        Py_DECREF(it);
+        return NULL;
+    }
+    bo->batch_size = n;
+    bo->it = it;
+    return (PyObject *)bo;
+}
+
+static void
+batched_dealloc(batchedobject *bo)
+{
+    PyObject_GC_UnTrack(bo);
+    Py_XDECREF(bo->it);
+    Py_TYPE(bo)->tp_free(bo);
+}
+
+static int
+batched_traverse(batchedobject *bo, visitproc visit, void *arg)
+{
+    if (bo->it != NULL) {
+        Py_VISIT(bo->it);
+    }
+    return 0;
+}
+
+static PyObject *
+batched_next(batchedobject *bo)
+{
+    Py_ssize_t i;
+    PyObject *it = bo->it;
+    PyObject *item;
+    PyObject *result;
+
+    // If the iterator has stopped, it should stay stopped
+    if (it == NULL) {
+        return NULL;
+    }
+    result = PyList_New(0);
+    if (result == NULL) {
+        return NULL;
+    }
+    for (i=0 ; i < bo->batch_size ; i++) {
+        item = (*Py_TYPE(it)->tp_iternext)(it);
+        if (item == NULL) {
+            break;
+        }
+        if (PyList_Append(result, item) < 0) {
+            Py_DECREF(item);
+            Py_DECREF(result);
+            return NULL;
+        }
+        Py_DECREF(item);
+    }
+    if (PyList_GET_SIZE(result) > 0) {
+        return result;
+    }
+    Py_CLEAR(bo->it);
+    Py_DECREF(result);
+    return NULL;
+}
+
+static PyTypeObject batched_type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    "itertools.batched",            /* tp_name */
```

Comment:
I won't insist on it but I feel the designated initializer syntax (`.tp_name = ...`) is a lot cleaner and safer. I understand if you don't want to change the existing types, but for new types it would be nice to use the new syntax.